### ISSUE TITLE
Weighting different training data

### ISSUE THREAD

**Issue by ps052537** [CompPhysVienna/n2p2#91]
I currently training the system with large numbers of training data. I am experiencing the bad fitness for the system with small number of atoms (molecules) than that with a large number of atoms.

I would like to increase weight of small molecules in the training data (weighing energy and forces instead of increasing the number of data), and does current version of n2p2 have this feature? if not, which part of the code should I check to add this feature?

**Comment by singraber** [comment]
Unfortunately a feature of weighing structures differently is not implemented. Does separate training of each set work as expected? Did you use the same reference method settings (e.g. ab initio k-points, ...) for both sets? These have to be consistent!

I can give you some guidelines how to add this feature but before I have to ask you to work out in more detail how the weighing should look in theory. If we're using this [publication](https://doi.org/10.1021/acs.jctc.8b01092) as basis, where would you add in the weighing factors?

**Comment by ps052537** [comment]
Can you descript how to seperate data set for training in a single training process? 
I am using the consistent reference method, and they are just different for the number of atoms. 

Weighing different dataset is conceptually simple. I remember that I can multiply weights on the value of loss of each data. Referring to the equation 22 in the paper you mentioned, it's multiplying different factors for different index a (for different training data) in the terms on the righthand side. (large factor for the small molecule in my case)

It would be great if I can apply this feacture so that get higher accuracy for smaller system.

**Comment by singraber** [comment]
I am not sure if I understood correctly your first paragraph... a test set is automatically kept aside from the entire data set when you use the `test_fraction` keyword. If you prefer to do that beforehand you can use the tool `nnp-select`.

Here is a rough sketch of the required changes to the source code for this weighing feature:

I would do this for each structure by adding it directly in the [input data file](https://compphysvienna.github.io/n2p2/topics/cfg_file.html). You could invent a new keyword (or use a special `comment` field), maybe like this:

```
begin
....
lattice ....
....
weight 1.5
atom .... 
energy ....
end
```
Then create an additional member variable `double weight` in the `Structure` class [(maybe after this line)](https://github.com/CompPhysVienna/n2p2/blob/3cfe391377d2792ac29baf8394b3dce712afdad2/src/libnnp/Structure.h#L84) to store the value

Next, we have to read in the weight factor when parsing the `input.data` file, this happens here:

https://github.com/CompPhysVienna/n2p2/blob/3cfe391377d2792ac29baf8394b3dce712afdad2/src/libnnp/Structure.cpp#L131

Just add something like this:

```cpp
        else if (splitLine.at(0) == "weight")
        {
            weight = atof(splitLine[1].c_str());
        }
```

Then during training you have to multiply the error and derivatives with this factor, this should go here:

errors:

https://github.com/CompPhysVienna/n2p2/blob/3cfe391377d2792ac29baf8394b3dce712afdad2/src/libnnptrain/Training.cpp#L2101-L2114

derivatives (energies):

https://github.com/CompPhysVienna/n2p2/blob/3cfe391377d2792ac29baf8394b3dce712afdad2/src/libnnptrain/Training.cpp#L1976-L1980

derivatives (forces):

https://github.com/CompPhysVienna/n2p2/blob/3cfe391377d2792ac29baf8394b3dce712afdad2/src/libnnptrain/Training.cpp#L2021-L2025

I think that should be everything... if you set up a feature branch I can help you if problems occur and eventually you can start also a pull request to add this feature into the main package. Would be really nice if you could contribute and test this feature :-)!

**Comment by ps052537** [comment]
Thanks so much for providing information about the points that I need to change! I made a new branch and let me add this feature.

**Comment by singraber** [comment]
Thinking again about this.. I forgot that one also needs to modify the "configuration sending/receiving" routines, otherwise the weights will not be transferred to other MPI tasks when running a parallel training:

At these two positions add lines in a similar fashion (actually just copy and exchange `volume` by `weight`:

https://github.com/CompPhysVienna/n2p2/blob/9c350fbb14f45ed4aaccf0dd3a691ded2a3d0be0/src/libnnptrain/Dataset.cpp#L277

https://github.com/CompPhysVienna/n2p2/blob/9c350fbb14f45ed4aaccf0dd3a691ded2a3d0be0/src/libnnptrain/Dataset.cpp#L494

Also, change 

https://github.com/CompPhysVienna/n2p2/blob/9c350fbb14f45ed4aaccf0dd3a691ded2a3d0be0/src/libnnptrain/Dataset.cpp#L182

to

```cpp
bs += 5 * cs + 4 * ss + 4 * is + 6 * ds;
```

**Comment by vsumaria** [comment]
Is there a way to think of adding weight to each atom in a structure separately? 

It's easy to I guess add that in the input.data and read that using the [Structure.cpp](https://github.com/CompPhysVienna/n2p2/blob/3cfe391377d2792ac29baf8394b3dce712afdad2/src/libnnp/Structure.cpp) code. 

If I am mainly interested in putting these weights to train forces on certain atoms better than the others, the energy error in training remains the same, but I would need to weight the force error vector according to these weights. How would one pass the atom weights to the structures? 

a.fRef[c->c] - a.f[c->c] --> this will then need to be multiplied with the specific weight for the 'a'
